File: libraries/node/src/sources/bms/convert.ts
/**
 * Mapper from the BMS-family source model onto a URC chart.
 */
import { UrcError } from "../../error.js";
import type { Chart, Layout, Metadata, Note, NoteType } from "../../model.js";
import { buildTiming, checkHoldOverlap, roundMs } from "../shared.js";
import {
	LAYOUTS,
	SIDE_OF,
	buildNotes,
	channelKind,
	detectMode,
	idValue
} from "./channels.js";
import type { BmsChart } from "./model.js";

const MEASURE_US = 240000000.0;
const TYPE_ORDER: Record<NoteType, number> = { N: 0, LS: 1, LE: 2, M: 3, F: 4 };

const SYSTEM_CHANNELS = new Set(["02", "03", "08", "09", "SC"]);

/**
 * Maps a BMS-family chart onto a URC chart.
 */
export function convertBms(chart: BmsChart): Chart {
	if (chart.bpm === null || chart.bpm <= 0)
		throw new UrcError("syntax", 1, "missing or non-positive #BPM");

	let maxMeasure = -1;
	for (const m of chart.measures.keys())
		if (m > maxMeasure)
			maxMeasure = m;

	const boundaries: number[] = [0.0];
	for (let m = 0; m <= maxMeasure; m++)
		boundaries.push(boundaries[boundaries.length - 1] + (chart.rates.get(m) ?? 1.0));

	type Entry = [number, number, "bpm" | "meter" | "stop" | "scroll" | "object" | "anchor", number];
	const entries: Entry[] = [[0.0, 0, "bpm", chart.bpm]];
	const objects: Array<[number, string, string]> = [];
	const used = new Set<string>();

	for (let m = 0; m <= maxMeasure; m++) {
		const rate = chart.rates.get(m) ?? 1.0;
		const prevRate = chart.rates.get(m - 1) ?? 1.0;
		if (rate !== prevRate) {
			const beats = rate * 4.0;
			if (Math.abs(beats - Math.round(beats)) < 1e-9 && Math.round(beats) >= 1)
				entries.push([boundaries[m], 3, "meter", Math.round(beats)]);
		}

		const measureMap = chart.measures.get(m);
		if (!measureMap)
			continue;

		for (const [channel, ids] of measureMap.entries())
			for (let idx = 0; idx < ids.length; idx++) {
				const obj = ids[idx];
				const y = boundaries[m] + (idx / ids.length) * rate;

				if (SYSTEM_CHANNELS.has(channel)) {
					if (obj === "00")
						continue;
					if (channel === "03") {
						const digits = idValue(obj, chart.base);
						entries.push([y, 0, "bpm", Math.floor(digits / 36) * 16 + (digits % 36)]);
					} else if (channel === "08") {
						const bpmVal = chart.bpmDefs.get(obj);
						if (bpmVal === undefined)
							throw new UrcError("syntax", 1, `undefined #BPM${obj}`);
						entries.push([y, 0, "bpm", bpmVal]);
					} else if (channel === "09") {
						const stopVal = chart.stopDefs.get(obj);
						if (stopVal === undefined)
							throw new UrcError("syntax", 1, `undefined #STOP${obj}`);
						entries.push([y, 1, "stop", stopVal]);
					} else {
						const scrollVal = chart.scrollDefs.get(obj);
						if (scrollVal === undefined)
							throw new UrcError("syntax", 1, `undefined #SCROLL${obj}`);
						entries.push([y, 2, "scroll", scrollVal]);
					}
					continue;
				}

				const kind = channelKind(channel);
				if (kind === null)
					continue;
				if (obj !== "00")
					used.add(`${SIDE_OF[channel[0]]}:${channel[1]}`);
				if (obj === "00" && kind !== "ln")
					continue;
				objects.push([y, channel, obj]);
				entries.push([y, 4, "object", objects.length - 1]);
			}
	}

	for (const y of boundaries)
		entries.push([y, 5, "anchor", 0]);

	const mode = detectMode(chart.pms, used);

	let bpm: number | null = null;
	let beats = 4;
	let timeUs = 0.0;
	let prevY = 0.0;
	let pendingStop = 0.0;
	const timed = new Array<number>(objects.length).fill(0.0);
	const bpmPoints: Array<[number, number, number, number]> = [];
	const svPoints: Array<[number, number]> = [];
	const anchors: number[] = [];

	entries.sort((a, b) => a[0] - b[0] || a[1] - b[1]);

	let i = 0;
	while (i < entries.length) {
		const y = entries[i][0];
		const group: Entry[] = [];
		while (i < entries.length && entries[i][0] === y) {
			group.push(entries[i]);
			i++;
		}

		if (bpm !== null)
			timeUs += (MEASURE_US * (y - prevY)) / bpm;
		timeUs += pendingStop;
		pendingStop = 0.0;

		let newBpm: number | null = bpm;
		let newBeats = beats;
		let scroll: number | null = null;

		for (const [, , kind, val] of group)
			if (kind === "bpm")
				newBpm = val;
			else if (kind === "meter")
				newBeats = val;
			else if (kind === "stop")
				pendingStop = (MEASURE_US * val) / newBpm!;
			else if (kind === "scroll")
				scroll = val;
			else if (kind === "anchor")
				anchors.push(roundMs(timeUs / 1000.0));
			else
				timed[val] = timeUs;

		if (newBpm !== bpm || newBeats !== beats)
			bpmPoints.push([roundMs(timeUs / 1000.0), newBpm!, newBeats, 4]);
		if (scroll !== null)
			svPoints.push([roundMs(timeUs / 1000.0), scroll]);

		bpm = newBpm;
		beats = newBeats;
		prevY = y;
	}

	const rawNotes = buildNotes(chart, mode, objects, timed);
	let firstNoteTime = 0;
	let hasFirst = false;
	for (const [t, , type] of rawNotes)
		if (type !== "LE")
			if (!hasFirst || t < firstNoteTime) {
				firstNoteTime = t;
				hasFirst = true;
			}

	const timing = buildTiming(
		bpmPoints,
		svPoints,
		firstNoteTime,
		".bms",
		anchors.find(time => time >= firstNoteTime) ?? null
	);

	const urcNotes: Note[] = rawNotes
		.map(([t, lane, type]) => ({
			timestampMs: t - firstNoteTime,
			lane,
			type
		}))
		.sort((a, b) => a.timestampMs - b.timestampMs || a.lane - b.lane || TYPE_ORDER[a.type] - TYPE_ORDER[b.type]);

	checkHoldOverlap(urcNotes);

	const [keys, specialKeys, specialLanes] = LAYOUTS[mode];
	const layout: Layout = {
		keys,
		specialKeys,
		specialLanes: specialLanes ? [...specialLanes] : null
	};

	const metadata: Metadata = {
		original: chart.pms ? "PMS" : "BMS",
		title: chart.title ?? "Unknown",
		artist: chart.artist ?? "Unknown",
		creator: "Unknown",
		version: chart.playLevel ?? "Unknown"
	};

	return {
		formatVersion: { major: 1, minor: 1 },
		metadata,
		judgment: null,
		layout,
		timing,
		notes: urcNotes
	};
}

File: libraries/rust/src/sources/bms/convert.rs
//! Mapper from the BMS-family source model onto a URC chart.

use std::collections::HashSet;

use crate::error::{Result, UrcError};
use crate::model::{Chart, Layout, Metadata, Note, NoteType, Version};
use crate::sources::shared::{build_timing, check_hold_overlap, round_ms};
use super::channels::{build_notes, channel_kind, detect_mode, id_value, resolve_layout, side_of};
use super::model::BmsChart;

const MEASURE_US: f64 = 240_000_000.0;

const SYSTEM_CHANNELS: [&str; 5] = ["02", "03", "08", "09", "SC"];

/// Maps a BMS-family chart onto a URC chart.
pub fn convert_bms(chart: &BmsChart) -> Result<Chart> {
    let bpm_initial = chart
        .bpm
        .ok_or_else(|| UrcError::new("syntax", 1, "missing or non-positive #BPM"))?;
    if bpm_initial <= 0.0 {
        return Err(UrcError::new("syntax", 1, "missing or non-positive #BPM"));
    }

    let max_measure = chart.measures.keys().copied().max().unwrap_or(-1);
    let mut boundaries = vec![0.0];
    for m in 0..=max_measure {
        let rate = chart.rates.get(&m).copied().unwrap_or(1.0);
        boundaries.push(boundaries.last().unwrap() + rate);
    }

    #[derive(Clone)]
    enum EntryKind {
        Bpm(f64),
        Meter(u64),
        Stop(f64),
        Scroll(f64),
        Object(usize),
        Anchor,
    }

    let mut entries: Vec<(f64, usize, EntryKind)> = vec![(0.0, 0, EntryKind::Bpm(bpm_initial))];
    let mut objects: Vec<(f64, String, String)> = Vec::new();
    let mut used = HashSet::new();

    for m in 0..=max_measure {
        let rate = chart.rates.get(&m).copied().unwrap_or(1.0);
        let prev_rate = chart.rates.get(&(m - 1)).copied().unwrap_or(1.0);
        if rate != prev_rate {
            let beats = rate * 4.0;
            if (beats - beats.round()).abs() < 1e-9 && beats.round() >= 1.0 {
                entries.push((
                    boundaries[m as usize],
                    3,
                    EntryKind::Meter(beats.round() as u64),
                ));
            }
        }

        if let Some(measure_map) = chart.measures.get(&m) {
            for (channel, ids) in measure_map {
                for (idx, obj) in ids.iter().enumerate() {
                    let y = boundaries[m as usize] + (idx as f64 / ids.len() as f64) * rate;

                    if SYSTEM_CHANNELS.contains(&channel.as_str()) {
                        if obj == "00" {
                            continue;
                        }
                        if channel == "03" {
                            let digits = id_value(obj, chart.base);
                            let bpm_val = ((digits / 36) * 16 + (digits % 36)) as f64;
                            entries.push((y, 0, EntryKind::Bpm(bpm_val)));
                        } else if channel == "08" {
                            let bpm_val = chart.bpm_defs.get(obj).copied().ok_or_else(|| {
                                UrcError::new("syntax", 1, format!("undefined #BPM{obj}"))
                            })?;
                            entries.push((y, 0, EntryKind::Bpm(bpm_val)));
                        } else if channel == "09" {
                            let stop_val = chart.stop_defs.get(obj).copied().ok_or_else(|| {
                                UrcError::new("syntax", 1, format!("undefined #STOP{obj}"))
                            })?;
                            entries.push((y, 1, EntryKind::Stop(stop_val)));
                        } else {
                            let scroll_val =
                                chart.scroll_defs.get(obj).copied().ok_or_else(|| {
                                    UrcError::new("syntax", 1, format!("undefined #SCROLL{obj}"))
                                })?;
                            entries.push((y, 2, EntryKind::Scroll(scroll_val)));
                        }
                        continue;
                    }

                    let kind = match channel_kind(channel) {
                        Some(k) => k,
                        None => continue,
                    };
                    if obj != "00" {
                        let side = side_of(channel.chars().next().unwrap());
                        let second = channel.chars().nth(1).unwrap();
                        used.insert((side, second));
                    }
                    if obj == "00" && kind != "ln" {
                        continue;
                    }

                    objects.push((y, channel.clone(), obj.clone()));
                    entries.push((y, 4, EntryKind::Object(objects.len() - 1)));
                }
            }
        }
    }

    for y in &boundaries {
        entries.push((*y, 5, EntryKind::Anchor));
    }

    let mode = detect_mode(chart.pms, &used);

    let mut bpm: Option<f64> = None;
    let mut beats = 4_u64;
    let mut time_us = 0.0;
    let mut prev_y = 0.0;
    let mut pending_stop = 0.0;
    let mut timed = vec![0.0; objects.len()];
    let mut bpm_points: Vec<(i64, f64, u64, u64)> = Vec::new();
    let mut sv_points: Vec<(i64, f64)> = Vec::new();
    let mut anchors: Vec<i64> = Vec::new();

    entries.sort_by(|a, b| {
        a.0.partial_cmp(&b.0)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| a.1.cmp(&b.1))
    });

    let mut i = 0;
    while i < entries.len() {
        let y = entries[i].0;
        let mut group = Vec::new();
        while i < entries.len() && entries[i].0 == y {
            group.push(entries[i].clone());
            i += 1;
        }

        if let Some(b) = bpm {
            time_us += MEASURE_US * (y - prev_y) / b;
        }
        time_us += pending_stop;
        pending_stop = 0.0;

        let mut new_bpm = bpm;
        let mut new_beats = beats;
        let mut scroll: Option<f64> = None;

        for (_, _, kind) in group {
            match kind {
                EntryKind::Bpm(v) => new_bpm = Some(v),
                EntryKind::Meter(v) => new_beats = v,
                EntryKind::Stop(v) => pending_stop = MEASURE_US * v / new_bpm.unwrap(),
                EntryKind::Scroll(v) => scroll = Some(v),
                EntryKind::Object(idx) => timed[idx] = time_us,
                EntryKind::Anchor => anchors.push(round_ms(time_us / 1000.0)),
            }
        }

        if new_bpm != bpm || new_beats != beats {
            bpm_points.push((round_ms(time_us / 1000.0), new_bpm.unwrap(), new_beats, 4));
        }
        if let Some(s) = scroll {
            sv_points.push((round_ms(time_us / 1000.0), s));
        }

        bpm = new_bpm;
        beats = new_beats;
        prev_y = y;
    }

    let raw_notes = build_notes(chart, mode, &objects, &timed)?;
    let first_note_time = raw_notes
        .iter()
        .filter(|(_, _, t)| *t != NoteType::Le)
        .map(|(time, _, _)| *time)
        .min()
        .unwrap_or(0);

    let anchor_ms = anchors
        .iter()
        .find(|time| **time >= first_note_time)
        .copied();

    let timing = build_timing(&bpm_points, &sv_points, first_note_time, ".bms", anchor_ms)?;

    let mut urc_notes: Vec<Note> = raw_notes
        .into_iter()
        .map(|(t, lane, note_type)| Note {
            timestamp_ms: t - first_note_time,
            lane,
            note_type,
        })
        .collect();

    fn type_order(nt: NoteType) -> usize {
        match nt {
            NoteType::N => 0,
            NoteType::Ls => 1,
            NoteType::Le => 2,
            NoteType::M => 3,
            NoteType::F => 4,
        }
    }

    urc_notes.sort_by(|a, b| {
        a.timestamp_ms
            .cmp(&b.timestamp_ms)
            .then_with(|| a.lane.cmp(&b.lane))
            .then_with(|| type_order(a.note_type).cmp(&type_order(b.note_type)))
    });

    check_hold_overlap(&urc_notes)?;

    let (keys, special_keys, special_lanes) = resolve_layout(mode);

    Ok(Chart {
        format_version: Version { major: 1, minor: 1 },
        metadata: Metadata {
            original: if chart.pms {
                "PMS".to_string()
            } else {
                "BMS".to_string()
            },
            title: chart.title.clone().unwrap_or_else(|| "Unknown".to_string()),
            artist: chart
                .artist
                .clone()
                .unwrap_or_else(|| "Unknown".to_string()),
            creator: "Unknown".to_string(),
            version: chart
                .play_level
                .clone()
                .unwrap_or_else(|| "Unknown".to_string()),
        },
        judgment: None,
        layout: Layout {
            keys,
            special_keys,
            special_lanes,
        },
        timing,
        notes: urc_notes,
    })
}



